### PR TITLE
TransactionBuilder TxOutContext

### PR DESCRIPTION
**Pull request by dolanbernard**
* Adding TxOutContext struct to return TxOut, TxOutConfirmationNumber, and a shared secret from transaction builder add_output
* Adding iOS native updates to expose TxOutContext to SDK
* Adding Android native updates to expose TxOutContext to SDK

### Motivation

This is needed for the SDKs to support Moby transaction history

### CODE REVIEW

File: transaction/std/src/transaction_builder.rs [hunk 1]
```diff
@@ -143,6 +159,26 @@ impl<FPR: FogPubkeyResolver> TransactionBuilder<FPR> {
...
+    /// Add a non-change output to the transaction.
+    ///
+    /// If a sender memo credential has been set, this will create an
+    /// authenticated sender memo for the TxOut. Otherwise the memo will be
+    /// unused.
+    ///
+    /// # Arguments
+    /// * `value` - The value of this output, in picoMOB.
+    /// * `recipient` - The recipient's public address
+    /// * `rng` - RNG used to generate blinding for commitment
```
Comment (iamalwaysuncomfortable) [inline transaction/std/src/transaction_builder.rs:175]: The documentation is exactly the same here as the preceeding method which is confusing given they're both public. Even if the differences are subtle, they should be explained

File: transaction/std/src/transaction_builder.rs [hunk 2]
```diff
@@ -143,6 +159,26 @@ impl<FPR: FogPubkeyResolver> TransactionBuilder<FPR> {
...
+    pub fn add_output_with_context<RNG: CryptoRng + RngCore>(
+        &mut self,
+        value: u64,
+        recipient: &PublicAddress,
+        rng: &mut RNG,
+    ) -> Result<TxOutContext, TxBuilderError> {
```
Comment (iamalwaysuncomfortable) [inline transaction/std/src/transaction_builder.rs:181]: I think this changes the transaction interface a little too much without much tangible benefit given the shared secret is possible to derive elsewhere. The ffis already compute the secret in this way here - a similar approach should probably be used for the memo decryption or if really needed to get the shared secret https://github.com/mobilecoinfoundation/mobilecoin/blob/9b08d3c0a6e5f4b866a8734e7c10274e404c056b/android-bindings/src/bindings.rs#L377

**Review by iamalwaysuncomfortable — CHANGES_REQUESTED** [review]
=

File: transaction/std/src/transaction_builder.rs [hunk 3]
```diff
@@ -38,6 +38,22 @@ impl TxOutputsOrdering for DefaultTxOutputsOrdering {
...
+pub struct TxOutContext {
+    /// TxOut that comes from a transaction builder
+    /// add_output/add_change_output
+    pub tx_out: TxOut,
+    /// confirmation that comes from a transaction builder
+    /// add_output/add_change_output
+    pub confirmation: TxOutConfirmationNumber,
+    /// Shared Secret that comes from a transaction builder
+    /// add_output/add_change_output
+    pub shared_secret: RistrettoPublic,
```
Comment (iamalwaysuncomfortable) [inline transaction/std/src/transaction_builder.rs:54]: This is a fairly sensitive piece of information, I'm not sure it should be coming out of transaction builder at all. The shared secret is derived via DH key exchange and (I think) can be derived on the other side/via ffi as long as the Moby app has the view private key. 

At the very least it shouldn't be a public field.

Comment (iamalwaysuncomfortable) [inline transaction/std/src/transaction_builder.rs:54]: This is a fairly sensitive piece of information, I'm not sure it should be coming out of transaction builder at all. The shared secret is derived via DH key exchange and can be derived on the other side/via ffi as long as the Moby app has the view private key. 

At the very least it shouldn't be a public field.

**Review by iamalwaysuncomfortable — APPROVED** [review]

**Review by iamalwaysuncomfortable — CHANGES_REQUESTED** [review]
I believe this is too large of a change to transaction builder & that we should probably not be sending the txout shared secret in a struct. Suggested some possible alternatives, would be happy to open up the discussion more.

File: transaction/std/src/transaction_builder.rs (hunk 3 above)
Comment (iamalwaysuncomfortable) [inline transaction/std/src/transaction_builder.rs:54]: This is a fairly sensitive piece of information, I'm not sure it should be coming out of transaction builder at all. The shared secret is derived via DH key exchange and can be derived on the other side/via ffi as long as the Moby app has the view private key.

At the very least it shouldn't be a public field.

Comment (garbageslam) [reply transaction/std/src/transaction_builder.rs:54]: @iamalwaysuncomfortable this is required for Moby to do their thing, they need to encrypt additional stuff using this field.

I also see nothing wrong with it being a public field, this can just be an organizational type like `InputSecret` or `OutputSecret` with no particular invariants

File: transaction/std/src/transaction_builder.rs [hunk 4]
```diff
@@ -38,6 +38,22 @@ impl TxOutputsOrdering for DefaultTxOutputsOrdering {
     }
 }
 
+/// Transaction output context is produced by add_output method
+/// Used for receipt creation
+#[allow(dead_code)]
```
Comment (eranrund) [inline transaction/std/src/transaction_builder.rs:43]: Why is this needed?

File: transaction/std/src/transaction_builder.rs (hunk 1 above)
Comment (eranrund) [reply transaction/std/src/transaction_builder.rs:175]: I'm wondering if it makes sense to keep both at all. I believe existing code could be changed to utilize struct deconstruction, e.g. `let TxOutContext { tx_out, confirmation, .. } = tx_builder.add_output(...);`

File: transaction/std/src/transaction_builder.rs (hunk 4 above)
Comment (dolanbernard) [reply transaction/std/src/transaction_builder.rs:43]: Without that compilation fails due to the shared secret being dead code.

**Comment by garbageslam** [comment]
I think this looks good in concept -- the only thing is, I think this should be the only version of `add_output` and `add_change_output`. That will lead to a large diff because there's a lot of code using the transaction builder, but I think that's okay. That's usually what we do when developing the transaction builder -- we allow breaking changes to happen to avoid at all costs having a confusing API

File: transaction/std/src/transaction_builder.rs (hunk 1 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/transaction_builder.rs:175]: Yeah I agree that this would be better

Comment (iamalwaysuncomfortable) [reply transaction/std/src/transaction_builder.rs:175]: Yeah if we're okay passing the shared secret as an output argument, this would be the better way

File: transaction/std/src/transaction_builder.rs (hunk 3 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/transaction_builder.rs:54]: My issue with passing the shared secret directly is Moby isn't the only consumer of the tx builder and exposing it for all possible clients here seems to break security.

Wouldn't in be possible to add a method within the ffis that generates the shared secret from the tx_out_public key? And then passes it upwards through the boundary? There are currently other ffi methods calculating the shared secret.

Comment (iamalwaysuncomfortable) [reply transaction/std/src/transaction_builder.rs:54]: Also, I'll have to go digest that MCIP

Comment (garbageslam) [reply transaction/std/src/transaction_builder.rs:54]: >  exposing it for all possible clients here seems to break security.

I guess I don't see why -- as long as they don't misuse it it's fine. The client already knows everything that this shared secret is concealing anyways.

I agree that this thing should be zeroized at the end but as long as that happens this isn't really a security issue. The clients compute the shared secret on their own for their own TxOut's all the time right now, it's part of the SDK interface.

Comment (iamalwaysuncomfortable) [reply transaction/std/src/transaction_builder.rs:54]: Yeah that's actually true, and James just pointed out it's in the same memory space. 

So: I'm sold. 

However, do we want to enforce that the TxBuilder add_output method now returns the shared secret as a tuple and the TxOutContext is constructed on the ffi's side via struct deconstruction? i.e. `let TxOutContext { tx_out, confirmation, .. } = tx_builder.add_output(...);`

File: transaction/std/src/transaction_builder.rs (hunk 2 above)
Comment (garbageslam) [reply transaction/std/src/transaction_builder.rs:181]: @iamalwaysuncomfortable that's not correct, the recipient of a TxOut is able to reconstruct the TxOut shared secret, but the sender is not, unless we expose either the tx out shared secret, or the tx private key, from the transaction builder. So they actually do need this or something like it if they want to encrypt additional messages using the shared secret.

File: transaction/std/src/transaction_builder.rs [hunk 5]
```diff
@@ -970,7 +989,7 @@ pub mod transaction_builder_tests {
                 get_input_credentials(block_version, amount, &sender, &fog_resolver, &mut rng);
             transaction_builder.add_input(input_credentials);
 
-            let (_txout, _confirmation) = transaction_builder
+            let _tx_out_context = transaction_builder
```
Comment (remoun) [inline transaction/std/src/transaction_builder.rs:992]:
```suggestion
            transaction_builder
```

File: libmobilecoin/src/transaction.rs [hunk 6]
```diff
@@ -481,6 +481,7 @@ pub extern "C" fn mc_transaction_builder_add_output(
     recipient_address: FfiRefPtr<McPublicAddress>,
     rng_callback: FfiOptMutPtr<McRngCallback>,
     out_tx_out_confirmation_number: FfiMutPtr<McMutableBuffer>,
+    out_tx_out_shared_secret: FfiMutPtr<McMutableBuffer>,
```
Comment (remoun) [inline libmobilecoin/src/transaction.rs:484]: Since we're switching `TransactionBuilder::add_*output` to return `TxOutContext`, would it make sense to similarly make these FFI `add_output` methods return the whole `TxOutContext`, replacing these 2 out pointers?

Comment (the-real-adammork) [reply libmobilecoin/src/transaction.rs:484]: yes, using the second out pointer was my idea, but I think returning a whole `TxOutContext` would be better.

Comment (the-real-adammork) [reply libmobilecoin/src/transaction.rs:484]: Ill work on a commit to make this change

**Review by garbageslam — APPROVED** [review]
LGTM

**Review by iamalwaysuncomfortable — APPROVED** [review]
Much better, cleared by me!

File: transaction/std/src/transaction_builder.rs (hunk 2 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/transaction_builder.rs]: Good clarification!

**Comment by remoun** [comment]
We should merge this PR/branch into `master`, too
